File: Shared/XOEmbedClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif
#if canImport(UIKit)
import UIKit
#endif

struct XOEmbedResponse: Decodable, Sendable {
    let url: URL
    let authorName: String
    let authorURL: URL?
    let html: String

    enum CodingKeys: String, CodingKey {
        case url
        case authorName = "author_name"
        case authorURL = "author_url"
        case html
    }
}

struct XOEmbedMetadata: Sendable {
    let canonicalURL: URL
    let authorName: String
    let postText: String?
}

enum XOEmbedError: LocalizedError {
    case invalidURL
    case invalidResponse
    case responseTooLarge
    case httpStatus(Int)

    var errorDescription: String? {
        switch self {
        case .invalidURL: String(localized: "error.invalid_x_url")
        case .invalidResponse: String(localized: "error.oembed_response")
        case .responseTooLarge: String(localized: "error.oembed_too_large")
        case .httpStatus(let status): String(localized: "error.oembed_http \(status)")
        }
    }
}

struct XOEmbedClient: Sendable {
    static let maximumResponseBytes = 512 * 1_024
    private let session: URLSession

    init(session: URLSession? = nil) {
        if let session {
            self.session = session
        } else {
            let configuration = URLSessionConfiguration.ephemeral
            configuration.timeoutIntervalForRequest = 12
            configuration.timeoutIntervalForResource = 15
            configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
            self.session = URLSession(configuration: configuration)
        }
    }

    func fetch(postURL: URL) async throws -> XOEmbedMetadata {
        guard let normalized = XURLValidator.normalizedPostURL(from: postURL),
              var components = URLComponents(string: "https://publish.x.com/oembed") else {
            throw XOEmbedError.invalidURL
        }
        components.queryItems = [URLQueryItem(name: "url", value: normalized.absoluteString)]
        guard let endpoint = components.url else { throw XOEmbedError.invalidURL }

        let (data, response) = try await session.data(from: endpoint)
        guard let response = response as? HTTPURLResponse else { throw XOEmbedError.invalidResponse }
        guard (200..<300).contains(response.statusCode) else { throw XOEmbedError.httpStatus(response.statusCode) }
        guard data.count <= Self.maximumResponseBytes else { throw XOEmbedError.responseTooLarge }

        let decoded = try JSONDecoder().decode(XOEmbedResponse.self, from: data)
        return XOEmbedMetadata(
            canonicalURL: XURLValidator.normalizedPostURL(from: decoded.url) ?? normalized,
            authorName: decoded.authorName,
            postText: Self.postText(from: decoded.html)
        )
    }

    static func postText(from html: String) -> String? {
        guard let expression = try? NSRegularExpression(
            pattern: #"<p(?:\s[^>]*)?>(.*?)</p>"#,
            options: [.caseInsensitive, .dotMatchesLineSeparators]
        ), let match = expression.firstMatch(
            in: html,
            range: NSRange(html.startIndex..., in: html)
        ), let range = Range(match.range(at: 1), in: html) else {
            return nil
        }
        let fragment = String(html[range])

        #if canImport(UIKit)
        guard let data = fragment.data(using: .utf8),
              let attributed = try? NSAttributedString(
                data: data,
                options: [.documentType: NSAttributedString.DocumentType.html, .characterEncoding: String.Encoding.utf8.rawValue],
                documentAttributes: nil
              ) else { return nil }
        let text = attributed.string.trimmingCharacters(in: .whitespacesAndNewlines)
        #else
        let text = fragment
            .replacingOccurrences(of: #"<[^>]+>"#, with: "", options: .regularExpression)
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&#39;", with: "'")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        #endif
        return text.isEmpty ? nil : text
    }
}

File: OshiLifeAndroid/lib/import/x_oembed_client.dart
import 'dart:convert';

import 'package:html_unescape/html_unescape.dart';
import 'package:http/http.dart' as http;
import 'package:oshilife/import/capped_http.dart';
import 'package:oshilife/import/x_url_validator.dart';

/// Port of `XOEmbedMetadata` (`Shared/Networking/XOEmbedClient.swift`).
class XOEmbedMetadata {
  const XOEmbedMetadata({
    required this.canonicalUrl,
    required this.authorName,
    this.postText,
    this.linkedUrls = const [],
  });

  final Uri canonicalUrl;
  final String authorName;
  final String? postText;
  final List<Uri> linkedUrls;
}

enum XOEmbedErrorKind {
  invalidUrl,
  invalidResponse,
  responseTooLarge,
  httpStatus,
}

/// Port of `XOEmbedError`. Carries the iOS localization key so the UI
/// layer can render the same localized message.
class XOEmbedException implements Exception {
  const XOEmbedException.invalidUrl()
    : kind = XOEmbedErrorKind.invalidUrl,
      statusCode = null;

  const XOEmbedException.invalidResponse()
    : kind = XOEmbedErrorKind.invalidResponse,
      statusCode = null;

  const XOEmbedException.responseTooLarge()
    : kind = XOEmbedErrorKind.responseTooLarge,
      statusCode = null;

  const XOEmbedException.httpStatus(int this.statusCode)
    : kind = XOEmbedErrorKind.httpStatus;

  final XOEmbedErrorKind kind;
  final int? statusCode;

  /// iOS `errorDescription` localization keys.
  String get localizationKey => switch (kind) {
    XOEmbedErrorKind.invalidUrl => 'error.invalid_x_url',
    XOEmbedErrorKind.invalidResponse => 'error.oembed_response',
    XOEmbedErrorKind.responseTooLarge => 'error.oembed_too_large',
    XOEmbedErrorKind.httpStatus => 'error.oembed_http',
  };

  @override
  String toString() => switch (kind) {
    XOEmbedErrorKind.httpStatus => 'XOEmbedException.httpStatus($statusCode)',
    _ => 'XOEmbedException.${kind.name}',
  };
}

/// Port of the `XOEmbedFetching` protocol (stub seam for the draft builder).
// ignore: one_member_abstracts
abstract interface class XOEmbedFetching {
  Future<XOEmbedMetadata> fetch(Uri postUrl);
}

/// Port of `XOEmbedClient` — the primary metadata source, using X's
/// official oEmbed endpoint (no API auth, no scraping).
class XOEmbedClient implements XOEmbedFetching {
  XOEmbedClient({http.Client? client}) : _client = client ?? http.Client();

  static const int maximumResponseBytes = 512 * 1024;

  final http.Client _client;

  static final RegExp _paragraphPattern = RegExp(
    r'<p(?:\s[^>]*)?>(.*?)</p>',
    caseSensitive: false,
    dotAll: true,
  );

  static final RegExp _anchorHrefPattern = RegExp(
    '''<a\\b[^>]*\\bhref\\s*=\\s*["']([^"']+)["']''',
    caseSensitive: false,
  );

  static final RegExp _brPattern = RegExp(r'<br\s*/?>', caseSensitive: false);

  static final RegExp _tagPattern = RegExp(r'<[^>]+>');

  @override
  Future<XOEmbedMetadata> fetch(Uri postUrl) async {
    final normalized = XUrlValidator.normalizedPostUrl(postUrl);
    if (normalized == null) throw const XOEmbedException.invalidUrl();

    final endpoint = Uri.https('publish.x.com', '/oembed', {
      'url': normalized.toString(),
    });

    final CappedResponse response;
    try {
      response = await cappedGet(
        _client,
        endpoint,
        maxBytes: maximumResponseBytes,
      );
    } on XOEmbedException {
      rethrow;
    } on http.ClientException {
      throw const XOEmbedException.invalidResponse();
    }
    if (!response.isSuccess) {
      throw XOEmbedException.httpStatus(response.statusCode);
    }
    if (response.exceededCap) {
      throw const XOEmbedException.responseTooLarge();
    }

    final Object? decoded;
    try {
      decoded = jsonDecode(utf8.decode(response.bodyBytes));
    } on FormatException {
      throw const XOEmbedException.invalidResponse();
    }
    if (decoded is! Map<String, dynamic>) {
      throw const XOEmbedException.invalidResponse();
    }
    final rawUrl = decoded['url'];
    final authorName = decoded['author_name'];
    final html = decoded['html'];
    if (rawUrl is! String || authorName is! String || html is! String) {
      throw const XOEmbedException.invalidResponse();
    }
    final decodedUrl = Uri.tryParse(rawUrl);
    if (decodedUrl == null) throw const XOEmbedException.invalidResponse();

    return XOEmbedMetadata(
      canonicalUrl: XUrlValidator.normalizedPostUrl(decodedUrl) ?? normalized,
      authorName: authorName,
      postText: postText(html),
      linkedUrls: linkedUrls(html),
    );
  }

  /// Port of `linkedURLs(from:)`: anchor hrefs with `&amp;` decoded,
  /// http(s) only, ordered dedup.
  static List<Uri> linkedUrls(String html) {
    final seen = <String>{};
    final results = <Uri>[];
    for (final match in _anchorHrefPattern.allMatches(html)) {
      final value = match.group(1)!.replaceAll('&amp;', '&');
      final url = Uri.tryParse(value);
      if (url == null) continue;
      final scheme = url.scheme.toLowerCase();
      if (scheme != 'http' && scheme != 'https') continue;
      if (!seen.add(url.toString())) continue;
      results.add(url);
    }
    return results;
  }

  /// Port of `postText(from:)`: isolates the first `<p>` paragraph and
  /// converts it to plain text — `<br>` becomes a newline, tags are
  /// stripped, and HTML entities are decoded (the iOS build does this via
  /// `NSAttributedString` HTML rendering; `html_unescape` reproduces the
  /// entity coverage).
  static String? postText(String html) {
    final match = _paragraphPattern.firstMatch(html);
    if (match == null) return null;
    final fragment = match.group(1)!;
    final text = HtmlUnescape()
        .convert(
          fragment.replaceAll(_brPattern, '\n').replaceAll(_tagPattern, ''),
        )
        .trim();
    return text.isEmpty ? null : text;
  }
}
